tests/postgresql/test-postgresql.cpp

```diff
@@ -1199,7 +1214,7 @@ struct test_enum_with_explicit_custom_type_int_rowset : table_creator_base
 
         try
         {
-            sql << "CREATE TABLE soci_test( Type smallint)";
+            sql << "CREATE TABLE soci_test( Type integer)";
```

Comment:
smallint is mapped to integer in the current code and int16 in the new one.

Reply:
@zann1x Does this mean that using `into(int_var)` from a column of `smallint` type doesn't work any longer? If so, this would be problematic as it still would be a silent break.

There is a discussion of allowing (at least broadening) implicit conversions in #1088 but can we do something like this for `into()` itself?

Reply:
That is basically what I meant [here](https://github.com/SOCI/soci/pull/954#issuecomment-1676396849):

> Note that users can still encounter `std::bad_cast` exceptions because of the stricter type mapping. This can result in the weird circumstance that the old API e.g. returns `dt_long_long`, which, despite the name, you now have to retrieve as `uint32_t` instead of `long long`. So far, this worked because of the `type_conversion`s in `unsigned-types.h`. Examples for that can be found in test-mysql.cpp ([1](https://github.com/SOCI/soci/pull/954/files#diff-3c7d5f688a808dc7d9b8b8e3cf70533fc76947df9a87c24a03ec963230b42782R703-R706), [2](https://github.com/SOCI/soci/pull/954/files#diff-3c7d5f688a808dc7d9b8b8e3cf70533fc76947df9a87c24a03ec963230b42782R715-R717), [3](https://github.com/SOCI/soci/pull/954/files#diff-3c7d5f688a808dc7d9b8b8e3cf70533fc76947df9a87c24a03ec963230b42782R726-R728)).
> We could possible relax the type check in `row::get(size_t)` in order to avoid some (or even all?) of the `std::bad_cast` exceptions if you're unhappy with these. The implicit type casts from #918 may be re-integrated here.

Reply:
#918 should be reworked as it declares template metaprog functions that are already standard in c++11.

Reply:
Adding to my previous answer: when using `smallint`, we would get a `std::bad_cast` exception in the test `test_enum_with_explicit_custom_type_int_rowset` (which uses `row`). The test `test_enum_with_explicit_custom_type_int_into` (which uses `into(...)`) works with both `smallint` and `integer`.

That's something I couldn't avoid for the `row` API. It's also something I wouldn't consider "silent" breakage because this always throws an exception instead of swallowing an unmapped/unhandled data type as it did before my changes in https://github.com/SOCI/soci/pull/954/commits/826fa7753a4a30c22e026847dec785aa133f0a3b.

It works for `into(...)` because `into` uses the enum `exchange_type`, which kind of has a dynamic type handling in its `standard_into_type` handlers already.
`row` on the other hand utilizes the class `type_holder` which stores a concrete type `T`. This concrete type is the type that we're mapping the database types to (via the enum `db_type`).

Reply:
So I guess we do need something along the lines of #1088 to preserve compatibility for row API uses (BTW, just to fix the definitions: "silent breakage" is code which continues to compile, without warnings, but changes behaviour, so this would be one).

@Sildra Would you agree with reworking your PR to only allow lossless implicit type conversions, i.e. those from smaller to bigger int types (of the same sign?), in the row API? If so, I still think we should merge this one and then merge your PR. TIA!

Reply:
I think I can allow this by comparing the sizeof of source and target as precheks and separating double. Will work on it tommorow.

Reply:
Better idea, I can change the static cast to perform runtime narrowing checks.

Reply:
https://github.com/Sildra/soci/commit/0eb519b71004cb71c1ff08ac9f84e3411df8ae4d

Let me know what you think of this implem.

Reply:
I'm not convinced that we need narrowing conventions, it risks being very confusing if your code works perfectly fine at first and then suddenly starts falling with an exception just because somebody added a value allowed by the column type but out of the C++ type range to the database.

But I'd also prefer to discuss this in its own PR, this one is becoming very difficult to navigate.